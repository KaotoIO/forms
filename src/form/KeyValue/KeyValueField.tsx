import { TextInputGroup, TextInputGroupMain } from '@patternfly/react-core';
import { forwardRef, useImperativeHandle, useRef } from 'react';
import { IDataTestID } from '../models';
import { useSuggestions } from '../hooks/suggestions';
import { JSONSchema4 } from 'json-schema';

interface KeyValueFieldProps extends IDataTestID {
  id: string;
  name: string;
  placeholder?: string;
  value?: string;
  onChange?: (value: string) => void;
  onFocus?: () => void;
  onBlur?: () => void;
}

const STRING_SCHEMA: JSONSchema4 = { type: 'string' };

export const KeyValueField = forwardRef<HTMLInputElement, KeyValueFieldProps>(
  ({ id, name, placeholder, value, 'data-testid': dataTestId, onChange, onFocus, onBlur }, ref) => {
    const inputRef = useRef<HTMLInputElement>(null);

    useImperativeHandle(ref, () => inputRef.current as HTMLInputElement);

    const suggestions = useSuggestions({
      propName: name,
      schema: STRING_SCHEMA,
      inputRef,
      value: value ?? '',
      setValue: onChange,
    });

    return (
      <TextInputGroup>
        <TextInputGroupMain
          type="text"
          id={id}
          name={name}
          data-testid={dataTestId}
          onChange={(_event, value) => {
            onChange?.(value);
          }}
          ref={inputRef}
          onFocus={onFocus}
          onBlur={onBlur}
          placeholder={placeholder}
          value={value}
        />

        {suggestions}
      </TextInputGroup>
    );
  },
);
