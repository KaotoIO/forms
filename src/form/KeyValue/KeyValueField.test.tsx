import { act, fireEvent, render, waitFor } from '@testing-library/react';
import { KeyValueField } from './KeyValueField';

// Mock useSuggestions to control its output
jest.mock('../hooks/suggestions', () => ({
  useSuggestions: jest.fn(() => null),
}));

describe('KeyValueField', () => {
  const defaultProps = {
    id: 'test-id',
    name: 'test-name',
    'data-testid': 'keyvalue-input',
    placeholder: 'Enter value',
    value: 'initial',
    onChange: jest.fn(),
    onFocus: jest.fn(),
    onBlur: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders with correct props', () => {
    const { getByTestId } = render(<KeyValueField {...defaultProps} />);
    const input = getByTestId('keyvalue-input');

    expect(input).toBeInTheDocument();
    expect(input).toMatchSnapshot();
  });

  it('calls onChange when value changes', () => {
    const onChange = jest.fn();
    const { getByRole } = render(<KeyValueField {...defaultProps} onChange={onChange} />);
    const input = getByRole('textbox');

    act(() => {
      fireEvent.change(input, { target: { value: 'new value' } });
    });

    expect(onChange).toHaveBeenCalledWith('new value');
  });

  it('should not fail if onChange was not provided', () => {
    const { getByRole } = render(<KeyValueField {...defaultProps} onChange={undefined} />);
    const input = getByRole('textbox');

    act(() => {
      expect(() => {
        fireEvent.change(input, { target: { value: 'new value' } });
      }).not.toThrow();
    });
  });

  it('calls onFocus and onBlur', async () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();

    const { getByRole } = render(<KeyValueField {...defaultProps} onFocus={onFocus} onBlur={onBlur} />);
    const input = getByRole('textbox');

    await act(async () => {
      fireEvent.focus(input);
      fireEvent.blur(input);
    });

    await waitFor(() => {
      expect(onFocus).toHaveBeenCalled();
      expect(onBlur).toHaveBeenCalled();
    });
  });

  it('shows suggestions when Ctrl+Space is pressed', async () => {
    const SuggestionsMenu = () => <div data-testid="suggestions-menu">Suggestions</div>;
    const { useSuggestions } = jest.requireMock('../hooks/suggestions');

    useSuggestions.mockImplementation(() => <SuggestionsMenu />);

    const { getByTestId } = render(<KeyValueField {...defaultProps} />);
    const input = getByTestId('keyvalue-input');

    await act(async () => {
      input.focus();
      fireEvent.keyDown(input, { code: 'Space', ctrlKey: true });
    });

    await waitFor(() => {
      expect(getByTestId('suggestions-menu')).toBeInTheDocument();
    });
  });
});
