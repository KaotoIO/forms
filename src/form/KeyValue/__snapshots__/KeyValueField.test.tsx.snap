// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyValueField renders with correct props 1`] = `
<div
  class="pf-v6-c-text-input-group__main"
  data-testid="keyvalue-input"
  id="test-id"
>
  <span
    class="pf-v6-c-text-input-group__text"
  >
    <input
      aria-label="Type to filter"
      class="pf-v6-c-text-input-group__text-input"
      name="test-name"
      placeholder="Enter value"
      type="text"
      value="initial"
    />
  </span>
</div>
`;
